pkg/trace/agent/agent.go

```diff
@@ -130,13 +134,21 @@ func (a *Agent) start() {
 	}
 }
 
-func (a *Agent) loop() {
-	ticker := time.NewTicker(a.conf.WatchdogInterval)
-	defer ticker.Stop()
+func (a *Agent) work() {
 	for {
 		select {
 		case t := <-a.Receiver.Out:
 			a.Process(t)
```

Comment:
Since `Process(t)` is now running in parallel would it be a good idea to get rid of the 3 separate goroutines it launches and just have it do all that work in serial right in `Process(t)`, then at the end block on writing to `a.tracePkgChan`? 

That way we have only one place where a large number of goroutines can get buffered right at the beginning of the agent pipeline, so it's a bit easier to reason about.

Reply:
I have briefly experimented with this as well, in benchmarks, and serial processing has proven to have a significant performance impact. What I would recommend, for now, is to stick with this minor change that has a potential for a big improvement, and re-assess other alternatives later. For example, another, perhaps even better alternative would be to indeed process inside each handler, and instead of having process workers, have separate workers for both the concentrator and sampling (which are now the two goroutines started by `Process`).

There are even more alternatives we can explore here, but I'd say for now to keep it simple and if it turns out to reveal problems, we can dig deeper and improve things. If we are to make changes, we'll need an actual investigation, profiles and benchmarks to prove that such a changes make sense.

Lastly, I am completely pro having the number of goroutines under control and starting 2 goroutines for each trace is something I'm not completely fond of either, and am honestly looking to exploring other ways.

Reply:
Sounds good 👍 This change will definitely help as is, and we can look into simplifying the downstream behavior later.